Update zh-tw.ctb to version 2017-11

school510587 opened the pull request:
Summary of changes:
1. Add \x2060. This looks like one space character.
2. Add missing subscript translations (\x2095 to \x209C).
3. Add \xF0B7. It is privately used for list item sign somewhere (such as
&nbsp;&nbsp;&nbsp;Microsoft Word).
4. Add -46 into \x2488-\x249B to indicate a full stop.
5. Assign the same dot pattern as "st" to both \xFB05 and \xFB06. Also add
&nbsp;&nbsp;&nbsp;\x017F as 234 (lowercase s).
6. Fix some Chinese words and Nemeth symbols brailled incorrectly.

Reviewed by:
&nbsp;&nbsp;&nbsp;&nbsp;Coscell Kao &lt;coscell@gmail.com&gt;
&nbsp;&nbsp;&nbsp;&nbsp;嘯傲俠羽 &lt;crazy@mail.batol.net&gt;

bertfrees commented:
@school510587 Thanks for the update!

I was wondering if you could help me with something? I'm trying to figure out what the differences are between all the Chinese tables that Liblouis contains. They are currently labelled as:
- zh-tw
- zh-hk (Chinese Hong Kong Cantonese braille Translation Table)
- zh-chn (Simplified-Chinese Braille Translation Table)

These names suggest that the first is applicable in Taiwan, the second in Hong Kong and the third in PRC. But I wonder if another subdivision would be better. Do any of these tables really need a region code? Of course, some languages and/or scripts are mainly used in specific regions, but if the braille rules for a specific language do not differ between regions, there is no reason to limit a table's use to a region.

So maybe zh-tw should just be called "cmn" (Mandarin), and zh-hk "yue" (Cantonese)?

On the other hand, if the braille rules apply to a specific region and are the same regardless of language used, labeling the braille table with a region tab is appropriate of course.

Braille rules could even be the same for all Chinese variants and for all regions? If that is the case there should be only one table.

zh-chn, which is a table for simplified Chinese, should maybe be called "zh-Hans"? Wikipedia tells me that in general mainland China uses simplified characters exclusively, while Hong Kong and Taiwan use traditional characters exclusively. But could someone in Taiwan translate a Mandarin text in simplified form to braille with the zh-chn table?

Another thing that confuses me is that the zh-chn table mentions Pinyin, which if I understand correctly is a romanization of Chinese.

school510587 commented:
Hi @bertfrees,

It is true that people in Taiwan, Hong Kong, and mainland learn different braille rules for the same Chinese. All of the tables are based on pronunciation, and they have nothing to do with how people write characters, i.e. traditional vs. simplified Chinese.

People in Taiwan mainly use ㄅㄆㄇㄈ (zhuyin, 注音), because students in elementary schools learn it. However, there is also a set of pinyin rules to express characters in Latin alphabets. Students are not necessary to learn these rules. I think it is mainly for communication with people whose native language is not Chinese. I am not familiar to the pinyin rules, too.

I do not know details about pronunciation system used in mainland, but I am sure that students there should learn to express every word in Latin alphabets. However, zh-tw.ctb and zh-chn.ctb have several differences. First, they use different dot patterns for the same pronunciation symbol. Second, they have quite different rules in displaying tonal marks. Our zh-tw.ctb requires a tonal mark at the last cell of each Chinese word, but zh-chn.ctb does not. These reasons make the two tables incompatible with each other.

Thus, the conclusion is that different braille rules are designed separately for different ways of denoting pronunciation. I really do not know why this phenomenon exists among people using the same language called Chinese. Due to this incompatibility, documents encoded in braille cannot be exchanged among Taiwan and mainland. We can only get the plaintext in Chinese characters and convert it into braille via a proper table according to what we have learned. Of course, a Taiwanese user may convert any Chinese characters using zh-chn.ctb, for special purposes such as teaching.

May the above answer help.

By the way, this also form an obstacle for Taiwanese users to read braille output of iPhones, because there is only rules used in PRC installed. However, many of my visually impaired friends use iPhones through their healthy ears. I am totally blind with poor hearing ability, so I have no chance to enjoy expensive iPhones. :)

school510587 commented:
Hi @bertfrees,

Please read these 3 pages:
1. https://en.wikipedia.org/wiki/Taiwanese_Braille
2. https://en.wikipedia.org/wiki/ Mainland_Chinese_Braille
3. https://en.wikipedia.org/wiki/Two-cell_Chinese_Braille

The first page is the set of rules that zh-tw.ctb follows.
The second page is the set of rules that zh-chn.ctb follows.
The third page is another set of rules used in mainland called two-cell braille, currently not implemented in any table of liblouis.
Further questions are welcome.

bertfrees commented:
Thanks so much for the clarification. That already helped me a lot.

So let me just get things straight. The following things are clear I think:

- Braille rules are based on region.
- Braille transcription is based entirely on pronunciation.
- Language is based more or less on region, however these regions are not exact countries. mainland and Taiwan do not exclusively speak Mandarin, and Hong Kong does not exclusively speak Cantonese.
- To use different writing systems (traditional vs. simplified) you need different unicode characters but also different fonts because some characters map to the same code points.

The following is not totally clear yet, but important to understand how a braille table works, since the transcription is based on pronunciation.

- How does pronunciation relate to language and writing system (traditional characters, simplified characters, pinyin, zhuyin)?

In the end, what I'm interested in is which Chinese tables can be used for which combinations of region, language and writing system.

According to the Wikipedia pages, mainland Chinese braille is used in PRC for standard Mandarin. In PRC, standard Mandarin is called Putonghua and is written in simplified form. But could mainland Chinese braille also be used for other Chinese languages and dialects like for instance Cantonese or Wu? Could a braille table convert from other scripts (traditional characters or zhuyin)? What does the braille table in Liblouis currently support?

Similarly, Wikipedia says Taiwanese braille is used in Taiwan for Taiwanese Mandarin (Guoyu), which is generally written in traditional characters. But could Taiwanese braille also be used for other Chinese languages, or could a table convert from simplified characters or from zhuyin? What does the braille table in Liblouis support?

The same questions apply for Cantonese braille and two-cell Chinese braille.

Could a braille table in theory convert pinyin to braille?

school510587 commented:
Hi @bertfrees, 

Now there are several writing, pronunciation, and braille systems used among Taiwan, Hong Kong and PRC.

&gt; How does pronunciation relate to language and writing system (traditional characters, simplified characters, pinyin, zhuyin)?
Here, languages are basically distinguished by pronunciation. Mandarin and Cantonese share exactly the same character set of all Chinese words and largely different rules of pronunciation. That is, selection of writing system is independent of selection of language.

&gt; What does the braille table in Liblouis currently support?
The key point is that the two writing systems are conceptually the same. That is, some traditional characters were "simplified" to construct a new writing system called simplified Chinese, but meaning of characters was unchanged. As a modern Chinese user, we should be ready to read both traditional and simplified Chinese everyday. zh-tw.ctb includes all of traditional and simplified characters. However, a user also cannot distinguish simplified characters from a mixture of both types of characters. This is also unimportant for Taiwanese users, because the semantic components are identical. When this is necessary, we can use character descriptions (eg. NVDA) or even the internal hexadecimal code of characters. The same principle also holds for zh-chn.ctb, zh-hk.ctb, and NVDA character descriptions of zh_CN, zh_HK and zh_TW. All of characterDescriptions.dic files of the three regions contain descriptions for both traditional and simplified Chinese characters.

After all, the conclusion is that the braille patterns have nothing to do with the writing systems. The same string written in "Chinese" can be always properly converted into Taiwanese braille, Hong Kong braille, or other two brailling rules (zh-chn.ctb and two-cell braille).

If this is not clear enough, further questions are welcome.

bertfrees commented:
OK, I think I'm starting to get the picture. Thanks :)

So the pronunciation, and therefore also the braille transcription, is independent from the writing system. This means that for a Liblouis table it does not matter which writing system is being used, however in order to support both, it must contain the Unicode characters of both traditional and simplified sets (which overlap to some extend but not completely). zh-tw.ctb supports both, zh-chn.ctb presumably not, and zh-hk.ctb probably also supports both but this needs to be double checked. Correct?

Mandarin and Cantonese have the same words but the pronunciation is different. Because braille transcription is based on pronunciation, this must mean that a Liblouis table for a certain braille system can only be used for the language it was designed for. So which braille system would a Cantonese-speaking user in mainland China learn (and which Liblouis table should be selected for that user)?

school510587 commented:
Hi @bertfrees,

Yes, zh-chn.ctb and zh-hk.ctb should contain both traditional and simplified characters for completeness of access to all Chinese documents. However, it is an alternative that text is processed by a traditional-to-simplified conversion program, but I think it is not the case. I don't know the actual set of characters they support, so it needs to be double checked.

I think a Cantonese-speaking user in mainland China may decide the table mainly based on his/her experience. Assume that this user was born in Hong Kong, so he/she learned to speak Cantonese before education. However, he/she was placed into an elementary school in mainland China where teachers used Mandarin brailling rules (zh-chn.ctb). Then, Mandarin brailling rules may be the most impressive for him/her, and he/she can speak both languages. Therefore, I think the process of selecting tables is also weakly related to which language a user learned. Of course, you must get more precise answers from real users in mainland China.

bertfrees commented:
OK, the Cantonese-speaking user in China was just an example. Maybe it was a bad one, but I certainly was not talking about immigrants. I thought Cantonese is not only spoken in Hong Kong, and Macau, but is also the language spoken in the whole province of Guangdong which is a huge area. I'm just getting this information from Wikipedia. I hope I'm not being ignorant!

What I'm trying to say is: in general, there are groups of people, in CHN, TW, HK or whatever, that don't speak the main language of their country.

Maybe in your country this is less the case, although [this website](https://www.ethnologue.com/country/TW/languages) lists no less than 25 languages for Taiwan. The second most common language is apparently [Taiwanese (Hokkien, Min Nan)](https://www.ethnologue.com/language/nan) that is spoken natively by about 70% of the population and is the de facto language of national identity. Maybe this is a somewhat controversial and geopolitical thing.

Anyway, what I'm basically trying to figure out is which braille systems these people that don't speak the official language of their country, use.

Thanks for all your time!

school510587 commented:
Hi @bertfrees,

There does exist some users in Taiwan can speak Hokkien only. However, they must either learn Mandarin and zhuyin (Taiwanese braille) or abandon braille functionalities of their screen readers. For young users, they will forced to learn Mandarin, the official language. For elder users, they usually abandon the braille.
It is definite that each way of pronunciation must have its dedicated brailling rules. I think the same case applies to other dialects spoken in Chinese world.

In fact, government in Taiwan has ever released some brailling rules for Hokkien. However, it is no popular, and users think it is too hard. In Taiwan, there is still no unified pronunciation representation for Hokkien. Complaint from users to the brailling rules therefore becomes an expectable result.

Do I hit the key point of your questions?

bertfrees commented:
Yes, you answered my question. Thanks!

So, my conclusion from all this is that the Liblouis tables should be labelled as follows:
- cmn-TW,
- cmn-CN or cmn-Hans-CN,
- yue-HK

The language tag "zh" should not be used because it represents a collection of several languages that all have different pronunciations and therefore must have dedicated braille systems.

Thanks for the discussion, I thought it was very interesting.

school510587 commented:
Hi @bertfrees,

You are welcome.

I think renaming zh tables is a great conclusion. After this, the table for two-cell braille of mainland China will be named properly. However, please notice that both current zh-chn.ctb and two-cell braille target Mandarin. That is, they are two different sets of rules for one pronunciation representation. Of course, you may treat two-cell braille as an improvement of the old rules currently listed in zh-chn.ctb. Their relationship is like that between UEB and the traditional grade-2 English braille. So, I think much more things should be considered before decision of new zh table labels.

There is another potential problems of data duplication. A ctb contains not only brailling rules of main character set used but also some symbols such as math operators. These symbols may be brailled the same in all variants of braille rules. In the future, you may receive requests that there should be an independent file to contain common rules.

bertfrees commented:
Yes, you're right. If we add a table for two-cell braille we'll need additional tags to distinguish the two cmn-CN, just like with UEB vs. EBAE. I'm planning to do that with a metadata field that will probably be called "system" and will contain a unique identifier of the implemented braille system. E.g.:
```
#+locale: cmn-CN
#+system: 2cell
```

But the locale is the most important piece of metadata today.

About the common rules: great, such changes are welcome.
